cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -439,6 +472,330 @@ struct ScalarMinMax {
   }
 };
 
+template <typename Type, typename Op>
+struct BinaryScalarMinMax {
+  using ArrayType = typename TypeTraits<Type>::ArrayType;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+  using offset_type = typename Type::offset_type;
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const ElementWiseAggregateOptions& options = MinMaxState::Get(ctx);
+    if (std::all_of(batch.values.begin(), batch.values.end(),
+                    [](const Datum& d) { return d.is_scalar(); })) {
+      return ExecOnlyScalar(ctx, options, batch, out);
+    }
+    return ExecContainingArrays(ctx, options, batch, out);
+  }
+
+  static Status ExecOnlyScalar(KernelContext* ctx,
+                               const ElementWiseAggregateOptions& options,
+                               const ExecBatch& batch, Datum* out) {
+    if (batch.values.empty()) {
+      return Status::OK();
+    }
+    auto output = checked_cast<BaseBinaryScalar*>(out->scalar().get());
+    if (!options.skip_nulls) {
+      // any nulls in the input will produce a null output
+      for (const auto& value : batch.values) {
+        if (!value.scalar()->is_valid) {
+          output->is_valid = false;
+          return Status::OK();
+        }
+      }
+    }
+    const Scalar& first_scalar = *batch.values.front().scalar();
+    string_view result = UnboxScalar<Type>::Unbox(first_scalar);
+    bool valid = first_scalar.is_valid;
+    for (size_t i = 1; i < batch.values.size(); i++) {
+      const Scalar& scalar = *batch[i].scalar();
+      if (!scalar.is_valid) {
+        DCHECK(options.skip_nulls);
```

Comment:
Why `DCHECK`? I think an `if` statement is better for this scenario.

Reply:
We take care of the case when the nulls aren't skipped and there are null values in the input right at the beginning of the method, so we assert it's impossible to have `options.skip_nulls` false here.